src/optimizer/rule/join_dependent_filter.cpp

```diff
@@ -0,0 +1,136 @@
+#include "duckdb/optimizer/rule/join_dependent_filter.hpp"
+
+#include "duckdb/optimizer/expression_rewriter.hpp"
+#include "duckdb/planner/expression/bound_columnref_expression.hpp"
+#include "duckdb/planner/expression/bound_comparison_expression.hpp"
+#include "duckdb/planner/expression/bound_conjunction_expression.hpp"
+#include "duckdb/planner/expression_iterator.hpp"
+
+namespace duckdb {
+
+JoinDependentFilterRule::JoinDependentFilterRule(ExpressionRewriter &rewriter) : Rule(rewriter) {
+	// Match on a ConjunctionExpression that has at least two ConjunctionExpressions as children
+	auto op = make_uniq<ConjunctionExpressionMatcher>();
+	op->matchers.push_back(make_uniq<ConjunctionExpressionMatcher>());
+	op->policy = SetMatcher::Policy::SOME;
+	root = std::move(op);
+}
+
+static inline void GetTableIndices(const Expression &expression, unordered_set<idx_t> &table_idxs) {
+	ExpressionIterator::EnumerateChildren(expression, [&](const Expression &child) {
+		if (child.GetExpressionClass() == ExpressionClass::BOUND_COLUMN_REF) {
+			auto &colref = child.Cast<BoundColumnRefExpression>();
+			table_idxs.insert(colref.binding.table_index);
+		} else {
+			GetTableIndices(child, table_idxs);
+		}
+	});
+}
+
+static inline bool ExpressionReferencesMultipleTables(const Expression &binding) {
+	unordered_set<idx_t> table_idxs;
+	GetTableIndices(binding, table_idxs);
+	return table_idxs.size() > 1;
+}
+
+static inline void ExtractConjunctedExpressions(Expression &expression,
+                                                unordered_map<idx_t, unique_ptr<Expression>> &expressions) {
+	if (expression.GetExpressionClass() == ExpressionClass::BOUND_CONJUNCTION &&
```

Comment:
nit, but this is not really necessary - `ExpressionType::CONJUNCTION_AND` implies `ExpressionClass::BOUND_CONJUNCTION`